GabijaBern/assignment2/BinaryTreeAncestors.java

```diff
@@ -0,0 +1,251 @@
+		if(nodeToFind  == null || givenTree == null)
+			return false;
+		
+		//base case - we found the node we have been looking, so we add it to the list
+		if(givenTree.treeRoot.value == nodeToFind.value)
+		{	
+			System.out.print("Node: " + givenTree.treeRoot.value + " ancestors: ");
+			return true;
+		}
+		
+		//climbing up from the base case and printing out the nodes
+		if(findAncestors(givenTree.treeRoot.left, nodeToFind)
+			|| findAncestors(givenTree.treeRoot.right, nodeToFind))
+		{
+			System.out.print(" " + givenTree.treeRoot.value);
+			return true;
+		}
+		
+		return false;
+	}
+	/*
+	 * For second part of the task we could edit the first method and make it find list of all nodes, rather 
+	 * than printing them out
+	 * 
+	 * @param givenTree The tree in which we will be searching
+	 * @param nodeToFind The node that we will be searching
+	 * @param list The list that we will be filling with the path from the node to the "main" root of the tree
+	 * 
+	 * 
+	 */
+	
+	public boolean findAncestors2(BTree<T> givenTree, Node<T> nodeToFind, List<Node<T>> list)
+	{
+		
+		//if we are given a null node or tree to search for/in
+		if(nodeToFind  == null || givenTree == null)
+			return false;
+		
+		//base case - we found the node we have been looking, so we add it to the list
+		if(givenTree.treeRoot.value == nodeToFind.value)
+		{	
+			list.add(givenTree.treeRoot);
+			return true;
+		}
+		
+		//climbing up the tree from the base case and adding the rest of the nodes
+		if(findAncestors2(givenTree.treeRoot.left, nodeToFind, list)
+			|| findAncestors2(givenTree.treeRoot.right, nodeToFind, list))
+		{
+			list.add(givenTree.treeRoot);
+			return true;
+		}
+		
+		return false;
+	}
+	
+	/*
+	 * Method for finding the closest common ancestor, since I am not overriding compareTo, all object are unique and if same values
+	 * appear in the tree, I don't have to worry, since I am comparing objects
+	 * 
+	 * @param tree The tree in which I should search for the nodes
+	 * @param firstNode The first node that we are going to search
+	 * @param secondNode The second node that we are going to search for and them find closest common ancestor for it and first one
+	 * 
+	 * @return Node<T> Method returns null if the tree doesn't exist, all other cases should be okay, since the main root can always
+	 * be the common ancestor
+	 * 
+	 * In this method I use edited version of first question part that finds list of "path" from one node to the "main" root
+	 * Then I find the intersection between two lists and first element in those lists should be the common ancestor
+	 */
+	public Node<T> findCommonAncestor(BTree<T> tree, Node<T> firstNode, Node<T> secondNode)
+	{
+		List<Node<T>> firstAncestors = new ArrayList<Node<T>>(); 
+		List<Node<T>> secondAncestors = new ArrayList<Node<T>>(); 
+		
+		findAncestors2(tree, firstNode, firstAncestors);
+		findAncestors2(tree, secondNode, secondAncestors);
+		
+		firstAncestors.retainAll(secondAncestors);

```

Comment:
Could you add a comment (or think about) time complexity here? I think this might be slower than it could be.